### DISCUSSION TITLE
Is it possible to createTheme in run time?

### DISCUSSION THREAD

**Discussion by hendry91** [facebook/stylex#276] [Q&A] [answered]
I'm currently working on a React project that requires the application of unique themes depending on the specific theme of each website. The goal is to dynamically retrieve theme configurations from Redux, or possibly create a custom hook that fetches these configurations from an external file. These themes then need to be applied at runtime using StyleX's createTheme function. Is such an implementation feasible? For example, if we manage 10 different websites, all under the same project, but each requiring a distinct theme, is it possible to inject these themes dynamically at runtime? Any insights or suggestions would be greatly appreciated. Thank you.

**Comment by hendry91** [comment]
```tsx
const stylesX = stylex.create({
    base: (theme) => ({
        fontSize: 16,
        background: theme.colors.secondary,
    }),
});

 const theme = useTheme();
<div className={styles.container} {...stylex.props(stylesX.base(theme))}></div>
```

I'm able to achieve what I want by using create props, wonder if createTheme can do the same?

**Comment by nmn** [comment, marked as answer]
> For example, if we manage 10 different websites, all under the same project, but each requiring a distinct theme

Let me start by solving your root problem. 

1. Use `stylex.defineVars` to define a set of variables that are shared by all 10 websites.
2. Create 10 different themes in 10 different files used `stylex.createTheme`.
3. Put the variables and themes in a common shared packages. (on NPM or monorepo)
4. Import the appropriate theme in each of the 10 websites.

Since only the styles from the files that are part of your bundle will be included in the generated CSS file, your problem is solved.

----------

Now, assuming there's more complexity involved, let me state a few additional things.

## If you want to dynamically select a theme at runtime

You can import all 10 themes in the same file and then conditionally apply only the one that is "active" on the page by using Redux, or a hook or whatever. Dynamically.

## If your theme itself is dynamic

If you don't know what your theme will be at all, and you're letting users use a color picker to choose colors and type in lengths etc, you can't use `createTheme`. However, you can use `stylex.create` with dynamic styles.

---------


```tsx
<div className={styles.container} {...stylex.props(stylesX.base(theme))}></div>
```

This is invalid. `styles.container` should be passed to `stylex.props`. Currently `stylex.props` will overrwrite the existing value of `className`.

Also, don't do this. You'll be setting inline styles everywhere and it'll be needlessly slow.

**Comment by hendry91** [comment]
Hi @nmn 
> You can import all 10 themes in the same file and then conditionally apply only the one that is "active" on the page by using Redux, or a hook or whatever. Dynamically.

10 themes is just an example, what if we have more than hundred website? I don't think importing all into the same file is ideal.

> If you don't know what your theme will be at all, and you're letting users use a color picker to choose colors and type in lengths etc, you can't use createTheme. However, you can use stylex.create with dynamic styles.

I'm a bit confused because stylex.create is only allowed at the root of a program and compiled at build time. we are not able to create it at runtime right? To summarize what I'm trying to achieve: let's say I have merchants A, B, C through Z, each with their own theme.json. When the website loads, I will to fetch the theme.json for the respective merchant and store it in Redux, with the aim of applying these themes to the components.

> this is invalid. styles.container should be passed to stylex.props. Currently stylex.props will overrwrite the existing value of className.
> Also, don't do this. You'll be setting inline styles everywhere and it'll be needlessly slow.

Sorry, Is the passing props method demonstrated below is recommended?

```tsx
const stylesX = stylex.create({
    base: (theme) => ({
        fontSize: 16,
        background: theme.colors.secondary,
    }),
});
 const theme = useTheme();
<div {...stylex.props(stylesX.base(theme))}></div>
```

Thanks..

**Reply by nmn** [reply]
I'll start with the answer you need:

```tsx
import {tokens} from './tokens.stylex'
const styles = stylex.create({
    createTheme: (values) => ({
        [tokens.primary]: theme.colors.primary,
        [tokens.secondary]: theme.colors.secondary,
        ...
    }),
});
 const theme = useTheme();
<div {...stylex.props(styles.createTheme(theme))}></div>
```

This will create a dynamic theme. It's not efficient, but it's better than using dyanamic styles everywhere. This way only this div is inefficient.

You can also use inline styles directly. `tokens.primary` is a string of the form `var(--foo)`, so you can drop the `var()` part and define values for the variables:

```tsx
import {tokens} from './tokens.stylex'
function getVar(var) {
  return var.slice(4, -1);
}

const theme = useTheme();
<div 
  style={{
    [getVar(tokens.primary)]: theme.colors.primary,
    [getVar(tokens.secondary)]: theme.colors.secondary,
    ...
  }}
/>
```
-----

What you're describing is that you want a single codebase to be able to have a hundred different themes where themes are loading at runtime.

Do you know the 100 themes while building the website? Could you build 100 different versions of the website where you import a different theme for each one?

That would be much more efficient.

**Comment by hendry91** [comment]
Hi @nmn 
The following code will encounter an error when processed by Lightningcss, do you have any idea?
```
import {tokens} from './tokens.stylex'
const styles = stylex.create({
    createTheme: (values) => ({
        [tokens.primary]: theme.colors.primary,
        [tokens.secondary]: theme.colors.secondary,
        ...
    }),
});
 const theme = useTheme();
<div {...stylex.props(styles.createTheme(theme))}></div>
```

![image](https://github.com/facebook/stylex/assets/15864186/a4452e45-3b7d-4a04-8454-a22f79edba1a)

**Reply by nmn** [reply]
This is a bug I'm tracking. Will fix soon.

**Reply by hendry91** [reply]
Cool.. Thanks @nmn

**Reply by nmn** [reply]
Just put up a PR.
